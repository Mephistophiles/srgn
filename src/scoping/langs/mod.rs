use super::Scoper;
#[cfg(doc)]
use crate::scoping::scope::Scope::{In, Out};
use crate::scoping::scope::{merge, subtract};
use log::{debug, trace};
use std::{ops::Range, str::FromStr};
pub use tree_sitter::{
    Language as TSLanguage, Parser as TSParser, Query as TSQuery, QueryCursor as TSQueryCursor,
};

/// C#.
pub mod csharp;
/// Python.
pub mod python;
/// TypeScript.
pub mod typescript;

/// Represents a (programming) language.
#[derive(Debug)]
pub struct Language<Q> {
    query: Q,
}

impl<Q> Language<Q> {
    /// Create a new language with the given associated query over it.
    pub fn new(query: Q) -> Self {
        Self { query }
    }
}

/// A query over a language, for scoping.
///
/// Parts hit by the query are [`In`] scope, parts not hit are [`Out`] of scope.
#[derive(Debug, Clone)]
pub enum CodeQuery<C, P>
where
    C: FromStr + Into<TSQuery>,
    P: Into<TSQuery>,
{
    /// A custom, user-defined query.
    Custom(C),
    /// A premade query.
    ///
    /// Availability depends on the language, respective languages features, and
    /// implementation in this crate.
    Premade(P),
}

impl<C, P> From<CodeQuery<C, P>> for TSQuery
where
    C: FromStr + Into<TSQuery>,
    P: Into<TSQuery>,
{
    fn from(value: CodeQuery<C, P>) -> Self {
        match value {
            CodeQuery::Custom(query) => query.into(),
            CodeQuery::Premade(query) => query.into(),
        }
    }
}

/// In a query, use this name to mark a capture to be ignored.
///
/// Useful for queries where tree-sitter doesn't natively support a fitting node type,
/// and a result is instead obtained by ignoring unwanted parts of bigger captures.
pub(super) const IGNORE: &str = "IGNORE";

/// A scoper for a language.
///
/// Functions much the same, but provides specific language-related functionality.
pub trait LanguageScoper: Scoper {
    /// The language's tree-sitter language.
    fn lang() -> TSLanguage;

    /// The language's tree-sitter query.
    fn query(&self) -> TSQuery;

    /// The language's tree-sitter parser.
    #[must_use]
    fn parser() -> TSParser {
        let mut parser = TSParser::new();
        parser
            .set_language(Self::lang())
            .expect("Should be able to load language grammar and parser");

        parser
    }

    /// Scope the given input using the language's query.
    ///
    /// In principle, this is the same as [`Scoper::scope`].
    fn scope_via_query(query: &mut TSQuery, input: &str) -> Vec<Range<usize>> {
        // tree-sitter is about incremental parsing, which we don't use here
        let old_tree = None;

        trace!("Parsing into AST: {:?}", input);

        let tree = Self::parser()
            .parse(input, old_tree)
            .expect("No language set in parser, or other unrecoverable error");

        let root = tree.root_node();
        debug!(
            "S expression of parsed source code is: {:?}",
            root.to_sexp()
        );

        let run = |query: &mut TSQuery| {
            trace!("Running query: {:?}", query);

            let mut qc = TSQueryCursor::new();
            let matches = qc.matches(query, root, input.as_bytes());

            let ranges = matches
                .flat_map(|query_match| query_match.captures)
                .map(|capture| capture.node.byte_range());

            let res = ranges.collect();
            trace!("Querying yielded ranges: {:?}", res);

            // Merge, because tree-sitter queries with multiple captures will return
            // them in some mixed order (not ordered, and not merged), but we later rely
            // on cleanly ordered, non-overlapping ranges (a bit unfortunate we have to
            // know about that remote part over here).
            merge(res)
        };

        let ranges = run(query);

        let has_ignore = query.capture_names().iter().any(|name| name == IGNORE);

        if has_ignore {
            let ignored_ranges = {
                disable_all_captures_except(IGNORE, query);

                debug!("Query has captures to ignore: running additional query");
                run(query)
            };

            let res = subtract(ranges, &ignored_ranges);
            debug!("Ranges cleaned up after subtracting ignores: {:?}", res);

            res
        } else {
            ranges
        }
    }
}

fn disable_all_captures_except(capture_name: &str, query: &mut TSQuery) {
    let capture_names = query.capture_names().to_owned();
    for name in capture_names {
        if name != capture_name {
            query.disable_capture(&name);
        }
    }
}
