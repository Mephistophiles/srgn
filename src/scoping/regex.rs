use regex::Regex;

use super::{langs::python::Scoper, ScopedView};

#[derive(Debug)]
pub struct RegexScoper {
    pattern: Regex,
    next: Option<Box<dyn Scoper>>,
}

impl RegexScoper {
    #[must_use]
    pub fn new(pattern: Regex, next: Option<Box<dyn Scoper>>) -> Self {
        Self { pattern, next }
    }
}

impl Scoper for RegexScoper {
    fn scope<'a>(&self, input: &'a str) -> ScopedView<'a> {
        let ranges = self.pattern.find_iter(input).map(|m| m.range());
        ScopedView::from_raw(input, ranges)
    }

    fn next(self) -> Option<Box<dyn Scoper>> {
        self.next
    }
}
