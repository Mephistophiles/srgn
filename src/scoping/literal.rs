use super::{ROScopes, Scoper};
use log::trace;
use std::{error::Error, fmt, ops::Range};
use unescape::unescape;

#[derive(Debug)]
pub struct Literal(String);

#[derive(Debug)]
pub enum LiteralError {
    InvalidEscapeSequences(String),
}

impl fmt::Display for LiteralError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidEscapeSequences(literal) => {
                write!(f, "Contains invalid escape sequences: '{literal}'")
            }
        }
    }
}

impl Error for LiteralError {}

impl TryFrom<String> for Literal {
    type Error = LiteralError;

    fn try_from(literal: String) -> Result<Self, Self::Error> {
        let unescaped =
            unescape(&literal).ok_or(LiteralError::InvalidEscapeSequences(literal.to_string()))?;

        Ok(Self(unescaped))
    }
}

impl Scoper for Literal {
    fn scope<'viewee>(&self, input: &'viewee str) -> ROScopes<'viewee> {
        let ranges = {
            let len = self.0.len();

            let ranges = input
                .match_indices(&self.0)
                .map(|(i, _)| Range {
                    start: i,
                    end: i + len,
                })
                .collect();

            trace!("Ranges in scope for {:?}: {:?}", self, ranges);

            ranges
        };

        ROScopes::from_raw_ranges(input, ranges)
    }
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use crate::scoping::{
        scope::{
            RWScope, RWScopes,
            Scope::{In, Out},
        },
        view::ScopedView,
    };
    use std::borrow::Cow::Borrowed;

    use super::*;

    #[rstest]
    #[case("a", "a", ScopedView::new(RWScopes(vec![RWScope(In(Borrowed("a")))])))]
    #[case("aa", "a", ScopedView::new(RWScopes(vec![RWScope(In(Borrowed("a"))), RWScope(In(Borrowed("a")))])))]
    #[case("aba", "a", ScopedView::new(RWScopes(vec![RWScope(In(Borrowed("a"))), RWScope(Out("b")), RWScope(In(Borrowed("a")))])))]
    //
    #[case(".", ".", ScopedView::new(RWScopes(vec![RWScope(In(Borrowed(".")))])))]
    #[case(r"\.", ".", ScopedView::new(RWScopes(vec![RWScope(Out(r"\")), RWScope(In(Borrowed(".")))])))]
    #[case(r".", r"\\.", ScopedView::new(RWScopes(vec![RWScope(Out(r"."))])))]
    //
    #[case("Hello\nWorld\n", "\n", ScopedView::new(RWScopes(vec![RWScope(Out("Hello")), RWScope(In(Borrowed("\n"))), RWScope(Out("World")), RWScope(In(Borrowed("\n")))])))]
    fn test_literal_scoping(
        #[case] input: &str,
        #[case] literal: &str,
        #[case] expected: ScopedView,
    ) {
        let builder = crate::scoping::view::ScopedViewBuilder::new(input);
        let literal = Literal::try_from(literal.to_owned()).unwrap();
        let actual = builder.explode(&literal).build();

        assert_eq!(actual, expected);
    }
}
