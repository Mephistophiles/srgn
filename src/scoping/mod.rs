//! Items for defining the scope actions are applied within.

use crate::actions::{self, Action};

use self::literal::LiteralError;
use self::regex::RegexError;
use itertools::Itertools;
use log::{debug, trace};
use std::fmt;
use std::{borrow::Cow, ops::Range};

pub mod langs;
pub mod literal;
pub mod regex;

#[derive(Debug)]
pub enum ScoperBuildError {
    EmptyScope,
    RegexError(RegexError),
    LiteralError(LiteralError),
}

impl From<LiteralError> for ScoperBuildError {
    fn from(e: LiteralError) -> Self {
        Self::LiteralError(e)
    }
}

impl From<RegexError> for ScoperBuildError {
    fn from(e: RegexError) -> Self {
        Self::RegexError(e)
    }
}

pub trait ScopedViewBuildStep {
    fn scope<'viewee>(&self, input: &'viewee str) -> ScopedViewBuilder<'viewee>;
}

impl fmt::Debug for dyn ScopedViewBuildStep {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Scoper").finish()
    }
}

/// Indicates whether a given string part is in scope.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Scope<'viewee, T> {
    /// The given part is in scope for processing.
    In(T),
    /// The given part is out of scope for processing.
    ///
    /// Treated as immutable, view-only.
    Out(&'viewee str),
}

type ROScope<'viewee> = Scope<'viewee, &'viewee str>;
type ROScopes<'viewee> = Vec<ROScope<'viewee>>;

type RWScope<'viewee> = Scope<'viewee, Cow<'viewee, str>>;
type RWScopes<'viewee> = Vec<RWScope<'viewee>>;

impl<'viewee> ROScope<'viewee> {
    /// Check whether the scope is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        let s: &str = self.into();
        s.is_empty()
    }
}

impl<'viewee> From<&'viewee ROScope<'viewee>> for &'viewee str {
    /// Get the underlying string slice of a [`ScopeStatus`].
    ///
    /// All variants contain such a slice, so this is a convenient method.
    fn from(s: &'viewee ROScope) -> Self {
        match s {
            Scope::In(s) | Scope::Out(s) => s,
        }
    }
}

impl<'viewee> From<ROScope<'viewee>> for RWScope<'viewee> {
    fn from(s: ROScope<'viewee>) -> Self {
        match s {
            Scope::In(s) => RWScope::In(Cow::Borrowed(s)),
            Scope::Out(s) => RWScope::Out(s),
        }
    }
}

impl<'viewee> From<&'viewee RWScope<'viewee>> for &'viewee str {
    /// Get the underlying string slice of a [`ScopeStatus`].
    ///
    /// All variants contain such a slice, so this is a convenient method.
    fn from(s: &'viewee RWScope) -> Self {
        match s {
            Scope::In(s) => s,
            Scope::Out(s) => s,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ScopedViewBuilder<'viewee> {
    scopes: ROScopes<'viewee>,
}

impl<'viewee> ScopedViewBuilder<'viewee> {
    #[must_use]
    pub fn new(input: &'viewee str) -> Self {
        Self {
            scopes: vec![Scope::In(input)],
        }
    }

    #[must_use]
    pub fn build(self) -> ScopedView<'viewee> {
        ScopedView {
            scopes: self
                .scopes
                .into_iter()
                .map(std::convert::Into::into)
                .collect(),
        }
    }
}

impl<'viewee> IntoIterator for ScopedViewBuilder<'viewee> {
    type Item = ROScope<'viewee>;

    type IntoIter = std::vec::IntoIter<ROScope<'viewee>>;

    fn into_iter(self) -> Self::IntoIter {
        self.scopes.into_iter()
    }
}

impl<'viewee> ScopedViewBuilder<'viewee> {
    #[must_use]
    pub fn explode_from_ranges(self, exploder: impl Fn(&str) -> Vec<Range<usize>>) -> Self {
        self.explode(|s| {
            trace!("Exploding from ranges: {:?}", s);

            let ranges = exploder(s);
            trace!("Raw ranges after exploding: {:?}", ranges);

            let mut scopes = Vec::new();

            let mut last_end = 0;
            for Range { start, end } in ranges.into_iter().sorted_by_key(|r| r.start) {
                scopes.push(Scope::Out(&s[last_end..start]));
                scopes.push(Scope::In(&s[start..end]));
                last_end = end;
            }

            if last_end < s.len() {
                scopes.push(Scope::Out(&s[last_end..]));
            }

            scopes.retain(|s| !s.is_empty());

            debug!("Scopes: {:?}", scopes);

            ScopedViewBuilder { scopes }
        })
    }

    #[must_use]
    pub fn explode_from_scoper(self, scoper: &impl ScopedViewBuildStep) -> Self {
        self.explode(|s| scoper.scope(s))
    }

    #[must_use]
    pub fn explode<F>(mut self, exploder: F) -> Self
    where
        F: Fn(&'viewee str) -> Self,
    {
        trace!("Exploding scopes: {:?}", self.scopes);
        let mut new = Vec::with_capacity(self.scopes.len());
        for scope in self.scopes.drain(..) {
            trace!("Exploding scope: {:?}", scope);

            if scope.is_empty() {
                trace!("Skipping empty scope");
                continue;
            }

            match scope {
                Scope::In(s) => {
                    let mut new_scopes = exploder(s).scopes;
                    new_scopes.retain(|s| !s.is_empty());
                    new.extend(new_scopes);
                }
                // Be explicit about the `Out(_)` case, so changing the enum is a
                // compile error
                Scope::Out("") => {}
                out @ Scope::Out(_) => new.push(out),
            }

            trace!("Exploded scope, new scopes are: {:?}", new);
        }
        trace!("Done exploding scopes.");

        ScopedViewBuilder { scopes: new }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ScopedView<'viewee> {
    scopes: RWScopes<'viewee>,
}

impl<'viewee> ScopedView<'viewee> {
    #[must_use]
    pub fn new(scopes: RWScopes<'viewee>) -> Self {
        Self { scopes }
    }

    /// For API discoverability.
    #[must_use]
    pub fn builder(input: &'viewee str) -> ScopedViewBuilder<'viewee> {
        ScopedViewBuilder::new(input)
    }

    /// submit a function to be applied to each in-scope, returning out-scopes unchanged
    pub fn map<F>(&mut self, f: &F) -> &mut Self
    where
        F: Fn(&str) -> <str as ToOwned>::Owned,
    {
        for scope in &mut self.scopes {
            match scope {
                Scope::In(s) => {
                    let res = f(s);
                    debug!(
                        "Replacing '{}' with '{}'",
                        s.escape_debug(),
                        res.escape_debug()
                    );
                    *scope = Scope::In(Cow::Owned(res));
                }
                Scope::Out(s) => {
                    debug!("Appending '{}'", s.escape_debug());
                }
            }
        }

        self
    }

    pub fn into_inner_mut(&mut self) -> &mut RWScopes<'viewee> {
        self.scopes.as_mut()
    }

    /// Check whether anything is in scope.
    #[must_use]
    pub fn has_any_in_scope(&self) -> bool {
        self.scopes.iter().any(|s| match s {
            Scope::In(_) => true,
            Scope::Out(_) => false,
        })
    }
}

/// Implementations of all available actions as dedicated methods.
///
/// Where actions don't take arguments, neither do the methods.
impl<'viewee> ScopedView<'viewee> {
    pub fn map_action<A: Action>(&mut self, action: &A) -> &mut Self {
        self.map(&|s| action.act(s))
    }

    pub fn delete(&mut self) -> &mut Self {
        let action = actions::Deletion::default();

        self.map_action(&action)
    }

    pub fn german(&mut self) -> &mut Self {
        let action = actions::German::default();

        self.map_action(&action)
    }

    pub fn lower(&mut self) -> &mut Self {
        let action = actions::Lower::default();

        self.map_action(&action)
    }

    pub fn normalize(&mut self) -> &mut Self {
        let action = actions::Normalization::default();

        self.map_action(&action)
    }

    pub fn replace(&mut self, replacement: String) -> &mut Self {
        let action = actions::Replacement::new(replacement);

        self.map_action(&action)
    }

    pub fn squeeze(&mut self) -> &mut Self {
        let action = actions::Squeeze::default();

        self.map_action(&action)
    }

    pub fn symbols(&mut self) -> &mut Self {
        let action = actions::Symbols::default();

        self.map_action(&action)
    }

    pub fn titlecase(&mut self) -> &mut Self {
        let action = actions::Titlecase::default();

        self.map_action(&action)
    }

    pub fn upper(&mut self) -> &mut Self {
        let action = actions::Upper::default();

        self.map_action(&action)
    }
}

impl fmt::Display for ScopedView<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for scope in &self.scopes {
            let s: &str = scope.into();
            write!(f, "{s}")?;
        }
        Ok(())
    }
}
