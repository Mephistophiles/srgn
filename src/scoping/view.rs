use crate::actions::{self, Action};
use crate::scoping::literal::LiteralError;
use crate::scoping::regex::RegexError;
use crate::scoping::scope::{ROScope, ROScopes, RWScope, RWScopes, Scope};
use crate::scoping::Scoper;
use log::{debug, trace};
use std::borrow::Cow;
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ScopedViewBuilder<'viewee> {
    scopes: ROScopes<'viewee>,
}

impl<'viewee> ScopedViewBuilder<'viewee> {
    #[must_use]
    pub fn new(input: &'viewee str) -> Self {
        Self {
            scopes: ROScopes(vec![ROScope(Scope::In(input))]),
        }
    }

    #[must_use]
    pub fn build(self) -> ScopedView<'viewee> {
        ScopedView {
            scopes: RWScopes(
                self.scopes
                    .0
                    .into_iter()
                    .map(std::convert::Into::into)
                    .collect(),
            ),
        }
    }
}

#[derive(Debug)]
pub enum ScoperBuildError {
    EmptyScope,
    RegexError(RegexError),
    LiteralError(LiteralError),
}

impl From<LiteralError> for ScoperBuildError {
    fn from(e: LiteralError) -> Self {
        Self::LiteralError(e)
    }
}

impl From<RegexError> for ScoperBuildError {
    fn from(e: RegexError) -> Self {
        Self::RegexError(e)
    }
}

impl<'viewee> IntoIterator for ScopedViewBuilder<'viewee> {
    type Item = ROScope<'viewee>;

    type IntoIter = std::vec::IntoIter<ROScope<'viewee>>;

    fn into_iter(self) -> Self::IntoIter {
        self.scopes.0.into_iter()
    }
}

impl<'viewee> ScopedViewBuilder<'viewee> {
    #[must_use]
    pub fn explode_from_scoper(self, scoper: &impl Scoper) -> Self {
        self.explode(|s| scoper.scope(s))
    }

    #[must_use]
    pub fn explode<F>(mut self, exploder: F) -> Self
    where
        F: Fn(&'viewee str) -> ROScopes<'viewee>,
    {
        trace!("Exploding scopes: {:?}", self.scopes);
        let mut new = Vec::with_capacity(self.scopes.0.len());
        for scope in self.scopes.0.drain(..) {
            trace!("Exploding scope: {:?}", scope);

            if scope.is_empty() {
                trace!("Skipping empty scope");
                continue;
            }

            match scope {
                ROScope(Scope::In(s)) => {
                    let mut new_scopes = exploder(s);
                    new_scopes.0.retain(|s| !s.is_empty());
                    new.extend(new_scopes.0);
                }
                // Be explicit about the `Out(_)` case, so changing the enum is a
                // compile error
                ROScope(Scope::Out("")) => {}
                out @ ROScope(Scope::Out(_)) => new.push(out),
            }

            trace!("Exploded scope, new scopes are: {:?}", new);
        }
        trace!("Done exploding scopes.");

        ScopedViewBuilder {
            scopes: ROScopes(new),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ScopedView<'viewee> {
    scopes: RWScopes<'viewee>,
}

/// Core implementations.
impl<'viewee> ScopedView<'viewee> {
    #[must_use]
    pub fn new(scopes: RWScopes<'viewee>) -> Self {
        Self { scopes }
    }

    /// For API discoverability.
    #[must_use]
    pub fn builder(input: &'viewee str) -> ScopedViewBuilder<'viewee> {
        ScopedViewBuilder::new(input)
    }

    /// Apply an action to all in-scope occurrences.
    ///
    /// See implementors of [`Action`] for available types.
    pub fn map(&mut self, action: &impl Action) -> &mut Self {
        for scope in &mut self.scopes.0 {
            match scope {
                RWScope(Scope::In(s)) => {
                    let res = action.act(s);
                    debug!(
                        "Replacing '{}' with '{}'",
                        s.escape_debug(),
                        res.escape_debug()
                    );
                    *scope = RWScope(Scope::In(Cow::Owned(res)));
                }
                RWScope(Scope::Out(s)) => {
                    debug!("Appending '{}'", s.escape_debug());
                }
            }
        }

        self
    }

    /// Squeeze all consecutive [`Scope::In`] scopes into a single occurrence (the first
    /// one).
    pub fn squeeze(&mut self) -> &mut Self {
        debug!("Squeezing view by collapsing all consecutive in-scope occurrences.");

        let mut prev_was_in = false;
        self.scopes.0.retain(|scope| {
            let keep = !(prev_was_in && matches!(scope, RWScope(Scope::In(_))));
            prev_was_in = matches!(scope, RWScope(Scope::In(_)));
            trace!("keep: {}, scope: {:?}", keep, scope);
            keep
        });

        debug!("Squeezed: {:?}", self.scopes);

        self
    }

    /// Check whether anything is in scope.
    #[must_use]
    pub fn has_any_in_scope(&self) -> bool {
        self.scopes.0.iter().any(|s| match s {
            RWScope(Scope::In(_)) => true,
            RWScope(Scope::Out(_)) => false,
        })
    }
}

/// Implementations of all available actions as dedicated methods.
///
/// Where actions don't take arguments, neither do the methods.
impl<'viewee> ScopedView<'viewee> {
    pub fn delete(&mut self) -> &mut Self {
        let action = actions::Deletion::default();

        self.map(&action)
    }

    #[cfg(feature = "german")]
    pub fn german(&mut self) -> &mut Self {
        let action = actions::German::default();

        self.map(&action)
    }

    pub fn lower(&mut self) -> &mut Self {
        let action = actions::Lower::default();

        self.map(&action)
    }

    pub fn normalize(&mut self) -> &mut Self {
        let action = actions::Normalization::default();

        self.map(&action)
    }

    pub fn replace(&mut self, replacement: String) -> &mut Self {
        let action = actions::Replacement::new(replacement);

        self.map(&action)
    }

    #[cfg(feature = "symbols")]
    pub fn symbols(&mut self) -> &mut Self {
        let action = actions::Symbols::default();

        self.map(&action)
    }

    #[cfg(feature = "symbols")]
    pub fn invert_symbols(&mut self) -> &mut Self {
        let action = actions::SymbolsInversion::default();

        self.map(&action)
    }

    pub fn titlecase(&mut self) -> &mut Self {
        let action = actions::Titlecase::default();

        self.map(&action)
    }

    pub fn upper(&mut self) -> &mut Self {
        let action = actions::Upper::default();

        self.map(&action)
    }
}

impl fmt::Display for ScopedView<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for scope in &self.scopes.0 {
            let s: &str = scope.into();
            write!(f, "{s}")?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::scoping::view::ScopedViewBuilder;
    use crate::RegexPattern;
    use rstest::rstest;

    #[rstest]
    // Pattern only
    #[case("a", "a", "a")]
    #[case("aa", "a", "a")]
    #[case("aaa", "a", "a")]
    //
    // Pattern once; nothing to squeeze
    #[case("aba", "a", "aba")]
    #[case("bab", "a", "bab")]
    #[case("babab", "a", "babab")]
    #[case("ababa", "a", "ababa")]
    //
    // Squeezes only the pattern, no other repetitions
    #[case("aaabbb", "a", "abbb")]
    //
    // Squeezes start
    #[case("aab", "a", "ab")]
    //
    // Squeezes middle
    #[case("baab", "a", "bab")]
    //
    // Squeezes end
    #[case("abaa", "a", "aba")]
    //
    // Squeezes as soon as pattern occurs at least twice
    #[case("a", "ab", "a")]
    #[case("ab", "ab", "ab")]
    #[case("aba", "ab", "aba")]
    #[case("abab", "ab", "ab")]
    #[case("ababa", "ab", "aba")]
    #[case("ababab", "ab", "ab")]
    //
    // Squeezes nothing if pattern not present
    #[case("", "b", "")]
    #[case("a", "b", "a")]
    #[case("aa", "b", "aa")]
    #[case("aaa", "b", "aaa")]
    //
    // Deals with character classes (space)
    #[case("Hello World", r"\s", "Hello World")]
    #[case("Hello  World", r"\s", "Hello World")]
    #[case("Hello       World", r"\s", "Hello World")]
    #[case("Hello\tWorld", r"\t", "Hello\tWorld")]
    #[case("Hello\t\tWorld", r"\t", "Hello\tWorld")]
    //
    // Deals with character classes (inverted space)
    #[case("Hello World", r"\S", "H W")]
    #[case("Hello\t\tWorld", r"\S", "H\t\tW")]
    //
    // Deals with overlapping matches; behavior of `regex` crate
    #[case("abab", r"aba", "abab")]
    #[case("ababa", r"aba", "ababa")]
    #[case("ababab", r"aba", "ababab")]
    #[case("abababa", r"aba", "abababa")]
    #[case("aba", r"aba", "aba")]
    #[case("abaaba", r"aba", "aba")]
    //
    // Requires non-greedy matches for meaningful results
    #[case("ab", r"\s+?", "ab")]
    #[case("a b", r"\s+?", "a b")]
    #[case("a\t\tb", r"\s+?", "a\tb")]
    #[case("a\t\t  b", r"\s+?", "a\tb")]
    //
    // Deals with more complex patterns
    #[case("ab", "", "ab")] // Matches nothing
    //
    #[case("ab", r"[ab]", "a")]
    #[case("ab", r"[ab]+", "ab")]
    #[case("ab", r"[ab]+?", "a")]
    //
    #[case("abab", r"\D", "a")]
    //
    // Builds up properly; need non-capturing group
    #[case("abab", r"(?:ab){2}", "abab")]
    #[case("ababa", r"(?:ab){2}", "ababa")]
    #[case("ababab", r"(?:ab){2}", "ababab")]
    #[case("abababa", r"(?:ab){2}", "abababa")]
    #[case("abababab", r"(?:ab){2}", "abab")]
    #[case("ababababa", r"(?:ab){2}", "ababa")]
    #[case("ababababab", r"(?:ab){2}", "ababab")]
    #[case("abababababab", r"(?:ab){2}", "abab")]
    //
    #[case("Anything whatsoever gets rEkT", r".", "A")]
    #[case(
    "Anything whatsoever gets rEkT",
    r".*", // Greediness inverted
    "Anything whatsoever gets rEkT"
)]
    //
    // Deals with Unicode shenanigans
    #[case("ðŸ˜ŽðŸ˜Ž", r"ðŸ˜Ž", "ðŸ˜Ž")]
    #[case("\0ðŸ˜Ž\0ðŸ˜Ž\0", r"ðŸ˜Ž", "\0ðŸ˜Ž\0ðŸ˜Ž\0")]
    //
    #[case("ä½ ä½ å¥½", r"ä½ ", "ä½ å¥½")]
    //
    // Longer ("integration") tests; things that come up in the wild
    #[case(
        " dirty Strings  \t with  \t\t messed up  whitespace\n\n\n",
        r"\s",
        " dirty Strings with messed up whitespace\n"
    )]
    #[case(
        " dirty Strings  \t with  \t\t messed up  whitespace\n\n\n",
        r" ",
        " dirty Strings \t with \t\t messed up whitespace\n\n\n"
    )]
    fn test_squeeze(#[case] input: &str, #[case] pattern: RegexPattern, #[case] expected: &str) {
        let builder = ScopedViewBuilder::new(input)
            .explode_from_scoper(&crate::scoping::regex::Regex::new(pattern.clone()));
        let mut view = builder.build();

        view.squeeze();
        let result = view.to_string();

        assert_eq!(result, expected);
    }
}
