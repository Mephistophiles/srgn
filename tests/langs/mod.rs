mod csharp;
mod go;
mod python;
mod rust;
mod typescript;

use srgn::scoping::{langs::LanguageScoper, regex::Regex, view::ScopedViewBuilder};
use std::{fs::read_to_string, path::Path};

fn get_input_output(lang: &str, file: &str) -> (String, String) {
    let path = Path::new("tests/langs");
    let path = path.join(lang);

    let input = read_to_string(path.join(format!("in/{file}"))).unwrap();
    let output = read_to_string(path.join(format!("out/{file}"))).unwrap();

    (input, output)
}

/// Nuke the target character from the input.
///
/// Convenience function for testing, as deleting a specific character, while
/// *retaining* it elsewhere, where the language did *not* scope down, is an easy way to
/// test.
fn nuke_target(input: &str, lang: &impl LanguageScoper) -> String {
    let mut builder = ScopedViewBuilder::new(input);

    builder.explode(lang);

    // Needs to be ASCII such that we can target e.g. variable names.
    let target = String::from("__T__");
    builder.explode(&Regex::try_from(target).unwrap());

    let mut view = builder.build();
    view.delete();

    view.to_string()
}
