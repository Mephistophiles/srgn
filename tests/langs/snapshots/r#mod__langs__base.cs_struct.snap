---
source: tests/langs/mod.rs
expression: inscope_parts
---
- n: 11
  l: "    public unsafe struct TestResult<T> where T : struct\n"
  m: "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^"
- n: 12
  l: "    {\n"
  m: ^^^^^^^
- n: 13
  l: "        public T Value;\n"
  m: ^^^^^^^^^^^^^^^^^^^^^^^^^
- n: 14
  l: "        public fixed byte ErrorCode[16];\n"
  m: ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- n: 15
  l: "    }\n"
  m: "^^^^^  "
